### PR TITLE
Add ability to pass multiple variables to Damaris using DamarisWriter class.

### PR DESCRIPTION
**Pull request by jcbowden**
The full list of available variables is hard coded and defined in initDamarisXmlFile.cpp ; 

Also adds ability to limit the variables to be passed through using the
```
 --damaris-limit-variables=<CSV list> 
```
command line option.

### CODE REVIEW

File: opm/simulators/flow/DamarisWriter.hpp [hunk 1]
```diff
@@ -154,6 +157,11 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
              "This name should be unique if multiple simulations are running on "
              "the same node/server as it is used for the Damaris shmem name and by "
              "the Python Dask library to locate sections of variables.");
+        Parameters::registerParam<TypeTag, Properties::DamarisLimitVariables>
+            ("A comma separated list of variable names that a user wants to pass"
+             "through via DamarisOutput::DamarisWriter::writeOutput)() to the "
+             "damaris_write() call. This can be used to limit the number of"
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:163]: Missing space after "of", and typo on next line "Daamis".

File: opm/simulators/flow/DamarisWriter.hpp [hunk 2]
```diff
@@ -259,15 +269,81 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
...
+                                                      this->elements_rank_offsets_[rank_]);
+            }
+
+            // Call damaris_write() for all available variables
+            for ( auto damVar : localCellData ) 
+            {
+               // std::map<std::string, data::CellData>
+              const std::string& name = damVar.first ;
+              
+              std::unordered_set<std::string>::const_iterator is_in_set = wanted_vars_set_.find ( name );
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:299]: Since this variable is only used to determine the presence in the set, I suggest instead changing the `if` on line 301 to `if (wanted_vars_set_.count(name)) {...`

File: opm/simulators/flow/DamarisWriter.hpp [hunk 3]
```diff
@@ -259,15 +269,81 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
...
+            }*/
+            
+            
+            // Call damaris_set_position() for all available variables
+            // There is an assumption that all variables are the same size, with the same offset.
+            // see initDamarisTemplateXmlFile.cpp for the Damaris XML descriptions.
+            for ( auto damVar : localCellData ) {
+                // std::map<std::string, data::CellData>
+                const std::string name = damVar.first ;
+                dam_err_ = DamarisOutput::setPosition(name.c_str(), rank_,
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:289]: Should this be called uncontitionally for all (possible) variables? Should we not do this inside the `if` below?

File: opm/simulators/flow/DamarisWriter.hpp [hunk 4]
```diff
@@ -259,15 +269,81 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
...
+               // std::map<std::string, data::CellData>
+              const std::string name = std::get<0>(damVar.first) ;
+              const int part = std::get<1>(damVar.first) ;
+              double  dataCol = damVar.second ;
+              std::cout << "Name of Damaris Block Varaiable : (" << rank_ << ")  "  << name  << "  part : " << part << "  Value : "  << dataCol <<  std::endl ;  
+            } 
+            
+            dam_err_ =  DamarisOutput::endIteration(rank_);
+            */
+            if (this->damarisOutputModule_->getPRESSURE_ptr() != nullptr)
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:342]: I do not understand what the logic is here? Also it seems a bit non-generic?

File: opm/simulators/flow/DamarisWriter.hpp [hunk 5]
```diff
@@ -259,15 +269,81 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
...
+            // Call damaris_write() for all available variables
+            for ( auto damVar : localCellData ) 
+            {
+               // std::map<std::string, data::CellData>
+              const std::string& name = damVar.first ;
+              
+              std::unordered_set<std::string>::const_iterator is_in_set = wanted_vars_set_.find ( name );
+              
+              if ((is_in_set != wanted_vars_set_.end() ) || (wanted_vars_set_.size() == 0)) {
+                  data::CellData  dataCol = damVar.second ;
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:302]: Is this copy necessary?

File: opm/simulators/flow/DamarisWriter.hpp [hunk 6]
```diff
@@ -259,15 +269,81 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
...
+                  OpmLog::debug(fmt::format("Name of Damaris Variable       : ( rank:{})  name: {}  ",  rank_, name));
+                  
+                  // It does not seem I can test for what type of data is present (double or int)
+                  // in the std::variant within the data::CellData, so I will use a try catch block. 
+                  // Although, only MPI_RANK and GLOBAL_CELL_INDEX are set as integer types (in the 
+                  // XML file) so it is a moot point. 
+                  // We could use damaris_get_type() to check what the type is specified as
+                  // within the Damaris XML file
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:310]: I see your point, usually std::holds_alternative could be used, but as the variant itself is not exposed that is not possible here. Maybe we should add an observer method to the data::CellData class? Alternatively, should we treat MPI_RANK and GLOBAL_CELL_INDEX specially to avoid this try-catch?

**Review by atgeirr — COMMENTED** [review]
I think this is a good way to dramatically improve the applicability of the Damaris layer. Some cleanup needed, and I have some questions, but this should get merged when finished.

File: opm/simulators/flow/DamarisWriter.hpp [hunk 7]
```diff
@@ -259,15 +269,81 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
...
                                                 this->damarisOutputModule_->getPRESSURE_ptr());
 
                 dam_err_ =  DamarisOutput::endIteration(rank_);
+            }*/
+            
+            
+            // Call damaris_set_position() for all available variables
+            // There is an assumption that all variables are the same size, with the same offset.
+            // see initDamarisTemplateXmlFile.cpp for the Damaris XML descriptions.
+            for ( auto damVar : localCellData ) {
```
[the remark this answers is not in the document]
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:286]: Yes, these are all read only, so can be constant, and references. My poor form really!

File: opm/simulators/flow/DamarisWriter.hpp (hunk 6 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:310]: Will there never be <int> values? If so, then using a variant seems redundant?

File: opm/simulators/flow/DamarisWriter.hpp (hunk 5 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:302]: Here, again, it's my lack of use of references.

File: opm/simulators/flow/DamarisWriter.hpp (hunk 4 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:342]: Me either. However, if this is not present, then the first iteration (iteration 0 in Damaris) is missed, and the GLOBAL_CELL_INDEX variable is not written. There is probably a better way to do this test, however, this works, and is derived from the original PRESSURE only code.

File: opm/simulators/flow/DamarisWriter.hpp (hunk 3 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:289]: Probably only needs to be done once (unless the MPI partitioning gets changed)

File: opm/simulators/flow/DamarisWriter.hpp (hunk 2 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:299]: Yes, I agree, you have a much nicer test.

**Comment by jcbowden** [comment]
I am still trying to test the Python interface better. Then on to Paraview.

**Comment by bska** [comment]
>  is there a way to obtain the output directory?

The `EclGenericWriter` base class has a protected `eclState_` data member of type `EclipseState`, so you can use a statement of the form
```C++
const auto& outputDir = this->eclState_.cfg().io().getOutputDir();
```
to get a read-only reference to a `std::string` containing the name of the output directory.  It should **probably** be a `std::filesystem::path` instead, but the `IOConfig` object is fairly old code by now and has not been made aware for the filesystem library.

**Comment by jcbowden** [comment]
Great, thanks. I am going to add the output path as a variable to the Damaris XML configuration file, so that Python can output data in the correct place. The Python interface does not have an 'outputdir' tag in its XML configuration, unlike HDF5, which has a 'FilesPath' tag.

**Comment by jcbowden** [comment]
Good afternoon, THe output directory information is available through the methods provided, however, when I am not using --output-dir="some/explicit/path/" then the output directory is defaulting to the directory of the input data. Is that the desired behavior?

**Comment by atgeirr** [comment]
> Good afternoon @atgeirr and @bska. The output directory information is available through the methods provided above, however, when I am not using --output-dir="some/explicit/path/" then the output directory is defaulting to the directory of the input data. Is that the desired behavior?

Yes it is. This is what the engineers and their workflows assume, so it is also our default.

File: opm/simulators/utils/DamarisKeywords.hpp [hunk 8]
```diff
@@ -104,6 +112,30 @@ getDamarisKeywords(const Parallel::Communication& comm, const std::string& Outpu
...
+    std::unordered_set<std::string> resuset ;
+    std::string tstr;
+    // The --damaris-limit-variables command line option (defaults to empty string)
+    std::string damarisLimitVars = Parameters::get<TypeTag, Properties::DamarisLimitVariables>();
+    std::stringstream ss(damarisLimitVars); 
+
+    // Use while loop to check the getline() function condition.  
+    while (std::getline(ss, tstr, ',')) {
+        //remove whitespace
+        // tstr.erase(std::remove_if(tstr.begin(), tstr.end(), ::isspace), tstr.end());
```
Comment (atgeirr) [inline opm/simulators/utils/DamarisKeywords.hpp:128]: Why did you change this to not use ::isspace?
If keeping current version, delete the commented out line.

**Comment by jcbowden** [comment]
Okay, I can see some utility in that. Anyway, I am almost ready to set this PR as "Ready for review" - I still need to go through some Paraview tests, and this will happen sometime soon.

File: opm/simulators/flow/DamarisWriter.hpp [hunk 9]
```diff
@@ -307,10 +390,18 @@ class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::G
         // We will add the MPI rank value directly into shared memory using the DamarisVar 
         // wrapper of the C based Damaris API.
         // The shared memory is given back to Damaris when the DamarisVarInt goes out of scope.
-        DamarisVarInt mpi_rank_var_test(1, {"n_elements_local"},  "MPI_RANK", rank_);
-        mpi_rank_var_test.setDamarisParameterAndShmem( {this->numElements_ } ) ;
+        // DamarisVarInt mpi_rank_var_test(1, {"n_elements_local"},  "MPI_RANK", rank_);
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.hpp:393]: Remove commented-out code.

File: opm/simulators/flow/DamarisWriter.hpp (hunk 4 above)
Comment (atgeirr) [reply opm/simulators/flow/DamarisWriter.hpp:338]: We can leave it like this for now.

File: opm/simulators/flow/DamarisWriter.hpp (hunk 6 above)
Comment (atgeirr) [reply opm/simulators/flow/DamarisWriter.hpp:297]: I do not understand the comment.

I suggest the following changes. First on line 297, add:
```
const bool data_is_integer = (name == "MPI_RANK" || name == "GLOBAL_CELL_INDEX");
```
Then instead of the try, use the `data_is_integer` variable to select which code block to run.

File: opm/simulators/flow/DamarisWriter.cpp [hunk 10]
```diff
@@ -42,7 +42,8 @@ int setPosition(const char* field, int rank, int64_t pos)
 {
     int dam_err = damaris_set_position(field, &pos);
     if (dam_err != DAMARIS_OK) {
-        OpmLog::error(fmt::format("damariswriter::writeOutput()       : ( rank:{})"
+        // OpmLog::error only prints from rank 0
+        OpmLog::error(fmt::format("damariswriter::setPosition()       : ( rank:{}) "
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.cpp:46]: I think we must use the DeferredLogger class here to ensure we get all error log messages. Then use gatherDeferredLogger() to gather them to rank 0 once we are in a context where every process will be. Tell me if you want some guidance on this.

Also: The `error` category should only be for things that stop the simulator. If this is not necessary (and I do not think it is, no exception is thrown and the process will keep running), use the `problem` (for numerical issues) or `warning` categories instead. Probably `warning` is the best here.

**Review by atgeirr — COMMENTED** [review]
Did a slightly more thorough review, and things are improving. Noticed some things that should be addressed, and think this can be merged when done.

File: opm/simulators/utils/DamarisKeywords.hpp (hunk 8 above)
Comment (jcbowden) [reply opm/simulators/utils/DamarisKeywords.hpp:128]: It did not work/compile - I forget the issue exactly.

File: opm/simulators/flow/DamarisWriter.cpp (hunk 10 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.cpp:46]: The test:
```
const bool data_is_integer = (name == "MPI_RANK" || name == "GLOBAL_CELL_INDEX");
```
 will not allow a future developer/user to add a variable that is of integer type. I assume that because a std::variant is used, that integer data vectors are possible.

Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.cpp:46]: hanged to warning. removed comments about only rank 0 getting the error/warning message - this still seems an issue as other ranks can have an error and the error message will be hidden.

File: opm/simulators/flow/DamarisWriter.hpp (hunk 6 above)
Comment (jcbowden) [reply opm/simulators/flow/DamarisWriter.hpp:297]: do you mean this comment: 
"Will there never be values? If so, then using a variant seems redundant?" or another one?
I think it may read better as:
 "Will there never be integer values? If so, then using a variant seems redundant?"

**Comment by jcbowden** [comment]
Setting this as ready for review - I am still testing the Damaris Paraview plugin, however this does not affect the opm-simulators.

**Comment by bska** [comment]
jenkins build this please

**Comment by jcbowden** [comment]
Hi, okay, so what is the issue? Just the compile time warnings?

**Comment by akva2** [comment]
if you are asking why the build is red, then yes that is due to the warnings.

**Comment by jcbowden** [comment]
Okay, i'll fix them. I have another update that removes the define compile time guards in DamarisKeywords.cpp. The guards affect the flexibility of changing the Damaris library capabilities without needing to re-compile opm-simulators. I will push this change with the warnings issues fix.

**Comment by jcbowden** [comment]
This does not have the int vs long int compiler issue fixed... give me a moment to think about it.

**Comment by bska** [comment]
jenkins build this please

**Comment by jcbowden** [comment]
A requirement is use of Damaris v.11.0 or higher as the HDF5 and Paraview plugins have been modified to work on iterations where no data is written to a particular variable (this seems common in flow). The Python plugin already tested for this, so earlier versions of Damaris should be okay if only using the Damaris Python output.

File: opm/simulators/flow/DamarisWriter.cpp [hunk 11]
```diff
@@ -42,7 +42,7 @@ int setPosition(const char* field, int rank, int64_t pos)
 {
     int dam_err = damaris_set_position(field, &pos);
     if (dam_err != DAMARIS_OK) {
-        OpmLog::error(fmt::format("damariswriter::writeOutput()       : ( rank:{})"
+        OpmLog::warning(fmt::format("damariswriter::setPosition()       : ( rank:{}) "
```
Comment (atgeirr) [inline opm/simulators/flow/DamarisWriter.cpp:45]: This will all be lost if rank is not 0. I can fix this in a followup, though.

**Comment by atgeirr** [comment]
Build is green, merging! Thanks for the effort.

**Comment by jcbowden** [comment]
Great! I had the the Norne case running with live Paraview output - the update now allows one to select from the (many) variables and have multiple variables all being viewed in 3D simultaneously. It worked quite well.
